### PR TITLE
lighttpd: update to lighttpd 1.4.69 release hash - backport to openwrt 21.02

### PR DESCRIPTION
**Pull request by gstrauss**
Maintainer: @flyn-org
Compile tested: arm_cortex-a9 OpenWrt master

Description:

update to lighttpd 1.4.69 release hash (from lighttpd 1.4.67)
includes backport of security bugfix to 21.02

Note: the commits are all straight cherry-picks of openwrt lighttpd commits on master with one merge conflict resolved to preserve meson build config, since openwrt 21.02 does not have meson build changes from (later) @neheb commit 5ccd4d1c250acb3d3e65f16e55926ef0bc80153d

Signed-off-by: Glenn Strauss [gstrauss@gluelogic.com](mailto:gstrauss@gluelogic.com)

### CODE REVIEW

**Comment by BKPepe** [comment]
I am somehow inclined to merge all these commits, well. My biggest doubt is that we are removing outdated/deprecated modules packages for users, which they could have installed, and the update at least for Turris routers using Turris OS would not proceed very well, and we will add more work for Turris/CZ.NIC employees even though, right now, they are slightly underpowered.

In my humble opinion, I would try to keep the changes as slim as possible to avoid bigger changes, which could affect the stable version and keep in mind, if we can not backport only the fix for security vulnerability or just simply bump it (which I don't think is possible due to modules, which are not simply there anymore, someone may correct me, if I am wrong.)

**Comment by 1715173329** [comment]
```console
2023-04-12T21:46:19.4332335Z make[2]: Entering directory '/feed/net/lighttpd'
2023-04-12T21:46:19.6956431Z mkdir -p /home/build/openwrt/dl
2023-04-12T21:46:19.7004178Z SHELL= flock /home/build/openwrt/tmp/.lighttpd-1.4.69.tar.xz.flock -c '  	/home/build/openwrt/scripts/download.pl "/home/build/openwrt/dl" "lighttpd-1.4.69.tar.xz" "16ac8db95e719629ba61949b99f8a26feba946a81d185215b28379bb4116b0b4" "" "https://download.lighttpd.net/lighttpd/releases-1.4.x"    '
2023-04-12T21:46:19.7616087Z rm -f /home/build/openwrt/build_dir/target-aarch64_cortex-a53_musl/lighttpd-1.4.69/.configured_*
2023-04-12T21:46:19.7639701Z rm -f /home/build/openwrt/staging_dir/target-aarch64_cortex-a53_musl/stamp/.lighttpd_installed
2023-04-12T21:46:19.7682207Z /home/build/openwrt/staging_dir/host/bin/sed -e "s|@CC@|'gcc',|" -e "s|@CXX@|'g++',|" -e "s|@PKGCONFIG@|/home/build/openwrt/staging_dir/host/bin/pkg-config|" -e "s|@CFLAGS@|'-O2', '-I/home/build/openwrt/staging_dir/host/include', '-I/home/build/openwrt/staging_dir/hostpkg/include', '-I/home/build/openwrt/staging_dir/target-aarch64_cortex-a53_musl/host/include', '-I/home/build/openwrt/staging_dir/host/include', '-I/home/build/openwrt/staging_dir/hostpkg/include', '-I/home/build/openwrt/staging_dir/target-aarch64_cortex-a53_musl/host/include',|" -e "s|@CXXFLAGS@|'-I/home/build/openwrt/staging_dir/host/include', '-I/home/build/openwrt/staging_dir/hostpkg/include', '-I/home/build/openwrt/staging_dir/target-aarch64_cortex-a53_musl/host/include',|" -e "s|@LDFLAGS@|'-L/home/build/openwrt/staging_dir/host/lib', '-L/home/build/openwrt/staging_dir/hostpkg/lib', '-L/home/build/openwrt/staging_dir/target-aarch64_cortex-a53_musl/host/lib',|" -e "s|@PREFIX@|/home/build/openwrt/staging_dir/hostpkg|" < /home/build/openwrt/staging_dir/hostpkg/lib/meson/openwrt-native.txt.in > /home/build/openwrt/build_dir/target-aarch64_cortex-a53_musl/lighttpd-1.4.69/openwrt-native.txt
2023-04-12T21:46:19.7703972Z bash: line 1: /home/build/openwrt/staging_dir/hostpkg/lib/meson/openwrt-native.txt.in: No such file or directory
2023-04-12T21:46:19.7709047Z make[2]: *** [Makefile:246: /home/build/openwrt/build_dir/target-aarch64_cortex-a53_musl/lighttpd-1.4.69/.configured_ca9776a3eb1721b014bc0d18bbdc88ed] Error 1
2023-04-12T21:46:19.7709535Z make[2]: Leaving directory '/feed/net/lighttpd'
2023-04-12T21:46:19.7713658Z time: package/feeds/packages_ci/lighttpd/compile#0.32#0.07#0.34
2023-04-12T21:46:19.7721786Z     ERROR: package/feeds/packages_ci/lighttpd failed to build.
2023-04-12T21:46:19.7728824Z make[1]: *** [package/Makefile:116: package/feeds/packages_ci/lighttpd/compile] Error 1
2023-04-12T21:46:19.7729332Z make[1]: Leaving directory '/home/build/openwrt'
2023-04-12T21:46:19.7733452Z make: *** [/home/build/openwrt/include/toplevel.mk:223: package/lighttpd/compile] Error 2
2023-04-12T21:46:21.4752726Z ##[error]Process completed with exit code 2.
```

**Comment by BKPepe** [comment]
Thank you so much, Glenn!

**Comment by gstrauss** [comment]
Modules now built-in to lighttpd are still created for compatibility.

However, a few long-deprecated modules were removed from the upstream source, and so those packages would no longer be available: lighttpd-mod-evasive, lighttpd-mod-secdownload, lighttpd-mod-uploadprogress, lighttpd-mod-usertrack.  I'll consider different options for this PR.

**Comment by gstrauss** [comment]
I added patches to restore the removed modules (for compatibility) and the build is in progress.  I am going to let the CI build proceed.  If the builds succeed, then I'll remove those patches and the cherry-picks for:

[lighttpd: remove deprecated modules](https://github.com/openwrt/packages/pull/20799/commits/17097d90fdb0960a2ff3cf6548268a94ff6a6a0f) [17097d9](https://github.com/openwrt/packages/pull/20799/commits/17097d90fdb0960a2ff3cf6548268a94ff6a6a0f)
[lighttpd: fix meson build](https://github.com/openwrt/packages/pull/20799/commits/dd57b9810e7c8e2b09a9bbc864485b77b7ee9680) [dd57b98](https://github.com/openwrt/packages/pull/20799/commits/dd57b9810e7c8e2b09a9bbc864485b77b7ee9680)

FYI: these long-deprecated modules were already removed last Nov in lighttpd-1.4.67-4 in openwrt-22.03 and in Turris OS 6.x, so no special changes are needed to those branches.

**Comment by gstrauss** [comment]
@1715173329  Can I get some help with this, please?  I built tools and toolchain on openwrt-21.02 branch and was able to build this PR successfully with `make V=s -j 4 -k package/feeds/packages/lighttpd/compile` after I renamed my prior existing `staging_dir/hostpkg` (via `mv staging_dir/hostpkg staging_dir/hostpkg.foo`) so that it would be recreated on the openwrt-21.02 branch.  I was unable to duplicate this error in my own build:
```
2023-04-14T07:49:02.7487496Z /home/build/openwrt/staging_dir/host/bin/sed -e "s|@CC@|'gcc',|" -e "s|@CXX@|'g++',|" -e "s|@PKGCONFIG@|/home/build/openwrt/staging_dir/host/bin/pkg-config|" -e "s|@CFLAGS@|'-O2', '-I/home/build/openwrt/staging_dir/host/include', '-I/home/build/openwrt/staging_dir/hostpkg/include', '-I/home/build/openwrt/staging_dir/target-powerpc_8540_musl/host/include', '-I/home/build/openwrt/staging_dir/host/include', '-I/home/build/openwrt/staging_dir/hostpkg/include', '-I/home/build/openwrt/staging_dir/target-powerpc_8540_musl/host/include',|" -e "s|@CXXFLAGS@|'-I/home/build/openwrt/staging_dir/host/include', '-I/home/build/openwrt/staging_dir/hostpkg/include', '-I/home/build/openwrt/staging_dir/target-powerpc_8540_musl/host/include',|" -e "s|@LDFLAGS@|'-L/home/build/openwrt/staging_dir/host/lib', '-L/home/build/openwrt/staging_dir/hostpkg/lib', '-L/home/build/openwrt/staging_dir/target-powerpc_8540_musl/host/lib',|" -e "s|@PREFIX@|/home/build/openwrt/staging_dir/hostpkg|" < /home/build/openwrt/staging_dir/hostpkg/lib/meson/openwrt-native.txt.in > /home/build/openwrt/build_dir/target-powerpc_8540_musl/lighttpd-1.4.69/openwrt-native.txt
2023-04-14T07:49:02.7503553Z bash: line 1: /home/build/openwrt/staging_dir/hostpkg/lib/meson/openwrt-native.txt.in: No such file or directory
```

**Comment by 1715173329** [comment]
1. I can reproduce this error on a clean buildroot, looks like `meson/host` is not built properly

2. fe018482f83d51b9005c44d25652ea323aa338a2 is also required, otherwise:
  `lighttpd-1.4.69/src/meson.build:271:0: ERROR: Dependency "libmaxminddb" not found, tried pkgconfig and cmake`

3. `libpcre2` was not selected properly
   ```
   Package lighttpd is missing dependencies for the following libraries:
   libpcre2-8.so.0
   ```

File: net/lighttpd/Makefile [hunk 1]
```diff
@@ -41,13 +86,15 @@ endef
 define Package/lighttpd
   $(call Package/lighttpd/Default)
   MENU:=1
-  DEPENDS:=+libnettle +libpcre2 +libpthread +LIGHTTPD_LOGROTATE:logrotate
+  DEPENDS:=+libpthread +LIGHTTPD_LOGROTATE:logrotate \
+           $(if $(CONFIG_LIGHTTPD_PCRE2),+libpcre2,) \
```
Comment (1715173329) [inline net/lighttpd/Makefile:90]: why not use `+LIGHTTPD_PCRE2:libpcre2` here?

**Review by 1715173329 — CHANGES_REQUESTED** [review]

File: net/lighttpd/Makefile [hunk 2]
```diff
@@ -41,13 +86,15 @@ endef
...
+           $(if $(cryptolib),+$(cryptolib),)
```
[the remark this answers is not in the document]
Comment (gstrauss) [reply net/lighttpd/Makefile]: Noted.

I added a commit with a minimal fix.  Other incorrect usage remains for subpackages, but they all depend on the lighttpd base package which happens to have a superset of the dependencies.  I'll look into fixing that on master with a future lighttpd release.

File: net/lighttpd/Makefile [hunk 3]
```diff
@@ -24,13 +24,58 @@ PKG_CPE_ID:=cpe:/a:lighttpd:lighttpd
...
-REBUILD_MODULES=authn_gssapi authn_ldap magnet webdav
+PKG_CONFIG_DEPENDS:= \
+  $(patsubst %,CONFIG_PACKAGE_lighttpd-mod-%,$(REBUILD_MODULES)) \
+  CONFIG_LIGHTTPD_PCRE2 \
+  CONFIG_LIGHTTPD_CRYPTOLIB_NONE \
+  CONFIG_LIGHTTPD_CRYPTOLIB_NETTLE \
+  CONFIG_LIGHTTPD_CRYPTOLIB_MBEDTLS \
+  CONFIG_LIGHTTPD_CRYPTOLIB_WOLFSSL
+
+PKG_BUILD_DEPENDS:= \
```
[the remark this answers is not in the document]
Comment (gstrauss) [reply net/lighttpd/Makefile:55]: Ick!  I had run into this too, but make a mistake on the merge.  Thanks for pointing it out.

File: net/lighttpd/Makefile (hunk 1 above)
Comment (gstrauss) [reply net/lighttpd/Makefile]: Will do.  I think I misunderstood the package caching from `make package/index` (before some of my make syntax can execute with config results) and make evaluation at build time.

**Comment by gstrauss** [comment]
@1715173329 thank you very much for the thorough review.  It has been extremely helpful to me.

**Comment by 1715173329** [comment]
In addition, please cherry-pick these commits for the maxminddb change:
- ce46bf8a4307ae2e0ec6d3f517cad05666eb7a22
- 508c4548dc7c73d6e824bd5a9b1dcb8fb7132ab3

Thank you!

**Review by 1715173329 — APPROVED** [review]
LGTM - Thanks!

**Comment by gstrauss** [comment]
6 architectures passed.
4 architectures failed, all with
```
Testing package knot-dig in version 3.2.4 from knot
Unknown package 'knot-dig'.
Collected errors:
 * pkg_hash_check_unresolved: cannot find dependency knot-libs (= 3.2.4-2) for knot-dig
 * pkg_hash_fetch_best_installation_candidate: Packages for knot-dig found, but incompatible with the architectures configured
 * opkg_install_cmd: Cannot install package knot-dig.
```

@1715173329 given that dependency is pre-existing in net/knot/Makefile, is there anything I need to do in this PR for knot or knot-dig?  Neither are related to lighttpd.

**Comment by 1715173329** [comment]
This is a known issue of CI, just ignore those errors ;)

**Comment by 1715173329** [comment]
... with a deeper review, I found more issues in current Makefile, but let's fix them in another PR ;)

**Comment by 1715173329** [comment]
Merged, thank you!

**Comment by gstrauss** [comment]
@1715173329 Thank you!  Follow-up fixes proposed in #20817 on master branch.

**Pull request merged**